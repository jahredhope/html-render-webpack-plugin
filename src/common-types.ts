import { Stats } from "webpack";
import MultiStats from "webpack/lib/MultiStats";
import { Request } from "express";

export type WebpackStats = Stats | MultiStats;
export type BaseRoute = { route: string };
export type Render<Route = BaseRoute> = (route: Route) => Promise<string>;
export type OnRendererReady<Route = BaseRoute> = (
  callback: (render: Render<Route>) => Promise<void>
) => void;

export type Renderer = (...params: any[]) => Promise<string>;
export type SourceModules = Record<string, string>;
export type RouteInput<Route = BaseRoute> = Route | string;
export type Trace = (...values: any[]) => void;
export type Specifier = string;
export type ExtraGlobals = object;
export type RenderConcurrency = "parallel" | "serial";
export type MapStatsToParams = ({
  webpackStats,
}: {
  webpackStats: WebpackStats;
}) => object;
export type TransformPath<Route = BaseRoute> = (route: Route) => string;
export type TransformExpressPath<Route = BaseRoute> = (route: Route) => string;
export type GetRouteFromRequest<Route> = (
  req: Request,
  routes: Route[]
) => Route;
