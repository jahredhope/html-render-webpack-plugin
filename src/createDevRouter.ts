import express, { Request, Response, NextFunction } from "express";
import { log, logError } from "./logging";
import exceptionFormatter from "exception-formatter";
import {
  OnRendererReady,
  BaseRoute,
  WebpackStats,
  TransformExpressPath,
  GetRouteFromRequest
} from "common-types";
import { Stats } from "webpack";

interface Params<Route> {
  routes: Route[];
  getRouteFromRequest?: GetRouteFromRequest<Route>;
  onRendererReady: OnRendererReady<Route>;
  transformExpressPath: TransformExpressPath<Route>;
  getClientStats: () => WebpackStats;
}

export = <Route extends BaseRoute>({
  routes,
  getRouteFromRequest,
  onRendererReady,
  transformExpressPath,
  getClientStats
}: Params<Route>) => {
  log("Create dev server");
  const formatErrorResponse = (
    error: string,
    webpackStats: Stats.ToJsonOutput
  ) => {
    let devServerScripts: string[] = [];
    if ("entrypoints" in webpackStats) {
      try {
        const devServerAssets = webpackStats.entrypoints!.main.assets;

        devServerScripts = devServerAssets.map(
          asset => `<script src="${webpackStats.publicPath}${asset}"></script>`
        );
      } catch (err) {
        console.error("Unable to load Dev Server Scripts. Error: ", err);
      }
    }

    return [error, ...devServerScripts].join("\n");
  };

  const devServerRouter = express.Router();

  const routesByExpressPath: Record<string, Route> = {};

  // Deduplicate paths to avoid duplicated processing in Express
  routes.forEach(route => {
    const expressPath = transformExpressPath(route);
    if (expressPath) {
      routesByExpressPath[expressPath] = route;
    }
  });

  Object.entries(routesByExpressPath).forEach(([expressPath, defaultRoute]) => {
    devServerRouter.get(
      expressPath,
      async (req: Request, res: Response, next: NextFunction) => {
        onRendererReady(async render => {
          const route = getRouteFromRequest
            ? getRouteFromRequest(req, routes)
            : defaultRoute;
          if (!route) {
            next();
          }
          if (!routes.includes(route)) {
            const errorMessage =
              "Returned route was not an existing route. Ensure return value from getRouteFromRequest is route";
            logError(errorMessage);
            throw new Error(errorMessage);
          }
          log(`Static render for ${route} from ${req.path}`);
          try {
            res.send(await render(route));
          } catch (error) {
            res.status(500).send(
              formatErrorResponse(
                exceptionFormatter(error, {
                  format: "html",
                  inlineStyle: true,
                  basepath: "webpack://static/./"
                }),
                getClientStats().toJson()
              )
            );
          }
        });
      }
    );
  });

  return devServerRouter;
};
