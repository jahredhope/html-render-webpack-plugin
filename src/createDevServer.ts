import express, { Request, Response } from "express";
import { log } from "./logging";
import exceptionFormatter from "exception-formatter";
import {
  OnRendererReady,
  BaseRoute,
  WebpackStats,
  TransformExpressPath
} from "common-types";
import { Stats } from "webpack";

interface Params<Route> {
  routes: Route[];
  onRendererReady: OnRendererReady<Route>;
  transformExpressPath: TransformExpressPath<Route>;
  getClientStats: () => WebpackStats;
}

export = <Route extends BaseRoute>({
  routes,
  onRendererReady,
  transformExpressPath,
  getClientStats
}: Params<Route>) => {
  log("Create dev server");
  const formatErrorResponse = (
    error: string,
    webpackStats: Stats.ToJsonOutput
  ) => {
    let devServerScripts: string[] = [];
    if ("entrypoints" in webpackStats) {
      try {
        const devServerAssets = webpackStats.entrypoints!.main.assets;

        devServerScripts = devServerAssets.map(
          asset => `<script src="${webpackStats.publicPath}${asset}"></script>`
        );
      } catch (err) {
        console.error("Unable to load Dev Server Scripts. Error: ", err);
      }
    }

    return [error, ...devServerScripts].join("\n");
  };

  const devServerRouter = express.Router();

  routes.forEach(route => {
    devServerRouter.get(
      transformExpressPath(route),
      async (req: Request, res: Response) => {
        onRendererReady(async render => {
          log(`Static render for ${route} from ${req.path}`);
          try {
            res.send(await render(route));
          } catch (error) {
            res.status(500).send(
              formatErrorResponse(
                exceptionFormatter(error, {
                  format: "html",
                  inlineStyle: true,
                  basepath: "webpack://static/./"
                }),
                getClientStats().toJson()
              )
            );
          }
        });
      }
    );
  });

  return devServerRouter;
};
