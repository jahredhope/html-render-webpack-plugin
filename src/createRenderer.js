const path = require("path");
const evaluate = require("eval");

function fileExistsInCompilation(specifier, compilation) {
  const fileName = path.basename(specifier);
  return Boolean(compilation.assets[fileName]);
}

function getFileSourceFromCompilation(specifier, compilation) {
  const fileName = path.basename(specifier);
  const asset = compilation.assets[fileName];
  return asset.source();
}

function evalutateFromSource(specifier, compilation, extraGlobals) {
  let source;
  try {
    source = getFileSourceFromCompilation(specifier, compilation);
  } catch (error) {
    throw new Error(`Error reading "${specifier}". Error: ${error}`);
  }
  return evaluate(
    source,
    /* filename: */ specifier,
    /* scope: */ {
      require: createLinker(specifier, compilation, extraGlobals),
      console,
      ...extraGlobals
    },
    /* includeGlobals: */ true
  );
}

function createLinker(parentModulePath, compilation, extraGlobals) {
  return function linker(specifier) {
    const absPath = path.join(path.dirname(parentModulePath), specifier);
    if (!fileExistsInCompilation(specifier, compilation)) {
      return require(specifier);
    }
    return evalutateFromSource(absPath, compilation, extraGlobals);
  };
}

module.exports = function createRenderer({
  fileName,
  renderCompilation,
  extraGlobals
}) {
  return evalutateFromSource(fileName, renderCompilation, extraGlobals);
};
