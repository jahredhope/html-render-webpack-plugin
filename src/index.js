const chalk = require("chalk");
const validateOptions = require("schema-utils");

const schema = require("./schemas/HtmlRenderWebpackPlugin.json");
const RenderError = require("./RenderError");
const renderHtml = require("./renderHtml");

const MultiStats = require("webpack/lib/MultiStats");

const returnEmptyObject = () => ({});
const defaultTransformFilePath = ({ route }) => route;

module.exports = class HtmlRenderPlugin {
  constructor(options = {}) {
    validateOptions(schema, options, "HTML Render Webpack Plugin");

    this.extraGlobals = options.extraGlobals || {};
    this.mapStatsToParams = options.mapStatsToParams || returnEmptyObject;
    this.renderDirectory = options.renderDirectory || "dist";
    this.renderEntry = options.renderEntry || "main";
    this.routes = options.routes || [""];
    this.transformFilePath =
      options.transformFilePath || defaultTransformFilePath;
    this.verbose = options.verbose || false;
    this.renderConcurrency = options.renderConcurrency || "serial";
    this.compilerToRenderWith = options.compilerToRenderWith;

    this.compilersComplete = 0;
    this.compilersRunning = 0;
    this.compilers = [];
    this.compilations = [];
    this.defaultHookOptions = { name: "HtmlRenderPlugin" };

    this.onRender = this.onRender.bind(this);
    this.logError = this.logError.bind(this);
    this.trace = this.trace.bind(this);
    this.apply = this.apply.bind(this);
    this.render = this.render.bind(this);
  }
  logError(...args) {
    console.log(chalk.red("ðŸš¨ HtmlRenderPlugin:"), ...args);
  }
  trace(...args) {
    if (this.verbose) {
      console.log(chalk.blue("HtmlRenderPlugin:"), ...args);
    }
  }
  async onRender(currentCompilation) {
    this.trace(`Starting render`);

    if (!this.renderCompiler) {
      const errorMessage = `Unable to find render compiler. Add \`.render()\` to any one configuration.`;
      this.logError(errorMessage);
      throw new Error(errorMessage);
    }
    if (!this.renderCompilation) {
      const errorMessage = `Unable to find render compilation. Something may have gone wrong during render build.`;
      this.logError(errorMessage);
      throw new Error(errorMessage);
    }

    const renderStats = this.renderCompilation.getStats();

    const webpackStats =
      this.compilations.length > 1
        ? new MultiStats(
            this.compilations.map(compilation => {
              return compilation.getStats();
            })
          )
        : currentCompilation.getStats();

    try {
      await renderHtml({
        extraGlobals: this.extraGlobals,
        mapStatsToParams: this.mapStatsToParams,
        renderConcurrency: this.renderConcurrency,
        renderCompilation: this.renderCompilation,
        renderCompiler: this.renderCompiler,
        renderDirectory: this.renderDirectory,
        renderEntry: this.renderEntry,
        renderStats: renderStats.toJson(),
        routes: this.routes,
        trace: this.trace,
        transformFilePath: this.transformFilePath,
        webpackStats
      });
    } catch (error) {
      this.logError("An error occured rendering HTML", error);
      currentCompilation.errors.push(new RenderError(error));
    }
  }
  applyRenderPlugin(compiler) {
    this.trace("Applying render plugin");
    this.renderCompiler = compiler;

    if (
      this.renderCompiler.options.output &&
      this.renderCompiler.options.output.path
    ) {
      this.renderDirectory =
        this.renderDirectory || this.renderCompiler.options.output.path;
    }
    this.renderCompiler.hooks.emit.tap(this.defaultHookOptions, compilation => {
      this.renderCompilation = compilation;
    });
  }
  applyCompiler(compiler, options = {}) {
    this.compilers.push(compiler);
    const compilerName = compiler.name || compiler.options.name;
    this.trace(`Recieved compiler: ${compilerName}`);

    compiler.hooks.run.tap(this.defaultHookOptions, () => {
      this.compilersRunning++;
    });
    compiler.hooks.watchRun.tap(
      this.defaultHookOptions,
      () => this.compilersRunning++
    );
    compiler.hooks.emit.tap(this.defaultHookOptions, compilation => {
      const index = this.compilers.indexOf(compilation.compiler);
      this.compilations[index] = compilation;
      this.compilersRunning--;
    });

    compiler.hooks.afterEmit.tapPromise(
      this.defaultHookOptions,
      async compilation => {
        this.compilersComplete++;
        if (this.compilersRunning > 0) {
          this.trace(
            `Assets emitted for ${compilerName}. Waiting for ${this.compilersRunning} other currently running compilers`
          );
          return;
        }
        if (this.compilersComplete < this.compilers.length) {
          this.trace(
            `Assets emitted for ${compilerName}. Waiting for ${this.compilers
              .length -
              this
                .compilersComplete} other compilers to finish their first build.`
          );
          return;
        }
        this.trace(
          `Assets emitted for ${compilerName}. compilersComplete ${this.compilersComplete}. No. Compilers: ${this.compilers.length}. Compilers running: ${this.compilersRunning}.`
        );
        return this.onRender(compilation);
      }
    );
    if (
      options.render ||
      (this.compilerToRenderWith && compilerName === this.compilerToRenderWith)
    ) {
      this.applyRenderPlugin(compiler);
    }
  }
  apply(compiler, options = {}) {
    const isMultiCompiler = Boolean(compiler.compilers);
    if (isMultiCompiler) {
      this.trace("Adding MultiCompiler");
      compiler.compilers.forEach(compiler => {
        const compilerName = compiler.name || compiler.options.name;
        const compilerToRenderWith = this.compilerToRenderWith || "render";
        this.applyCompiler(compiler, {
          ...options,
          render: compilerToRenderWith === compilerName
        });
      });
    } else {
      this.trace("Adding Compiler");
      this.applyCompiler(compiler, options);
    }
  }
  render() {
    return compiler => this.apply(compiler, { render: true });
  }
};
