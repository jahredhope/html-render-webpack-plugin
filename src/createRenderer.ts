import evalutateFromSource from "./evalutateFromSource";
import { SourceModules, ExtraGlobals } from "common-types";

export = function createRenderer({
  fileName,
  source,
  extraGlobals
}: {
  fileName: string;
  source: SourceModules;
  extraGlobals: ExtraGlobals;
}) {
  if (!fileName) {
    throw new Error("Missing filename");
  }
  if (!source) {
    throw new Error("Missing source");
  }
  return evalutateFromSource(fileName, source, extraGlobals);
};
