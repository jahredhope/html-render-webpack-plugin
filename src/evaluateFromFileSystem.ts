import path from "path";
import evaluate from "eval";
import { ExtraGlobals, FileSystem } from "./common-types";

import { log } from "./logging";

function getFromSourceModules(
  specifier: string,
  fsModule: FileSystem,
  rootDir: string
) {
  const sourceModuleSpecifier = specifier.replace(/^\.\//, "");
  if (!fsModule.existsSync(path.resolve(rootDir, sourceModuleSpecifier))) {
    log(
      `Unable to find file specifier ${sourceModuleSpecifier}. Root: ${rootDir}.`
    );
    return undefined;
  }
  return fsModule.readFileSync(path.resolve(rootDir, sourceModuleSpecifier));
}

function evaluateFromFileSystem(
  specifier: string,
  fsModule: FileSystem,
  rootDir: string,
  extraGlobals: ExtraGlobals
) {
  log(
    `Evaluating source for ${specifier}". From root directory: "${rootDir}".`
  );
  let source;
  try {
    source = getFromSourceModules(specifier, fsModule, rootDir);
  } catch (error) {
    throw new Error(`An error reading "${specifier}". Error: ${error}`);
  }
  return evaluate(
    source,
    /* filename: */ specifier,
    /* scope: */ {
      console,
      process,
      ...(extraGlobals || {}),
      require: createLinker(specifier, fsModule, rootDir, extraGlobals),
    },
    /* includeGlobals: */ true
  );
}

function createLinker(
  parentModulePath: string,
  fsModule: FileSystem,
  rootDir: string,
  extraGlobals: ExtraGlobals
) {
  log("Creating linker for", parentModulePath);
  return function linker(specifier: string) {
    const absPath = path.join(path.dirname(parentModulePath), specifier);
    if (!getFromSourceModules(specifier, fsModule, rootDir)) {
      log(`Using external require for ${specifier} from ${parentModulePath}`);
      return require(specifier);
    }
    log(`Linking ${parentModulePath} to asset ${specifier}`);
    return evaluateFromFileSystem(absPath, fsModule, rootDir, extraGlobals);
  };
}

export = evaluateFromFileSystem;
