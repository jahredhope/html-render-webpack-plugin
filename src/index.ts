import validateOptions from "schema-utils";

import schema from "./schemas/HtmlRenderWebpackPlugin.json";
import RenderError from "./RenderError";
import renderRoutes from "./renderRoutes";
import { log, logError } from "./logging";
import MultiStats from "webpack/lib/MultiStats";
import createDevRouter from "./createDevRouter";
import {
  BaseRoute,
  ExtraGlobals,
  MapStatsToParams,
  RenderConcurrency,
  Renderer,
  Render,
  RouteInput,
  OnRendererReady,
  TransformExpressPath,
  TransformPath,
  WebpackStats,
  GetRouteFromRequest,
} from "./common-types";
import { Compiler, compilation, Stats } from "webpack";
import getSourceFromCompilation from "./getSourceFromCompilation";
import createRenderer from "./createRenderer";
import { Router } from "express";

const timeSince = (startTime: number) => `${(Date.now() - startTime) / 1000}s`;
const defaultMapStats = ({ webpackStats }: { webpackStats: WebpackStats }) =>
  webpackStats ? { webpackStats: webpackStats.toJson() } : {};
const defaultTransform: TransformPath = <Route extends BaseRoute>(
  route: Route
) => route.route;

interface Options<Route extends BaseRoute = BaseRoute> {
  skipAssets?: boolean;
  getRouteFromRequest?: GetRouteFromRequest<Route>;
  routes?: RouteInput<Route>[];
  mapStatsToParams?: MapStatsToParams;
  renderDirectory?: string;
  renderConcurrency?: RenderConcurrency;
  transformFilePath?: TransformPath<Route>;
  transformExpressPath?: TransformExpressPath<Route>;
  renderEntry?: string;
  extraGlobals?: ExtraGlobals;
}

interface CompilationStatus {
  compilation: compilation.Compilation | null;
  isReady: boolean;
}

export = class HtmlRenderPlugin<Route extends BaseRoute = BaseRoute> {
  constructor(options: Options<Route> = {}) {
    validateOptions(schema, options || {}, "HTML Render Webpack Plugin");

    const pluginName = "HtmlRenderPlugin";

    const {
      extraGlobals = {},
      skipAssets = false,
      mapStatsToParams = defaultMapStats,
      renderEntry = "main",
      getRouteFromRequest,
      transformFilePath = defaultTransform,
      transformExpressPath = defaultTransform,
      renderConcurrency = "serial",
    } = options;

    const routes: Route[] = (options.routes || [""]).map((route) =>
      typeof route === "string" ? ({ route } as Route) : route
    );

    const renderDirectory = options.renderDirectory || "dist";

    const clientCompilations: CompilationStatus[] = [];
    let rendererCompilation: CompilationStatus;

    let renderer: Renderer;
    let lastClientStats: MultiStats | Stats | null = null;

    const isBuildReady = () =>
      rendererCompilation &&
      rendererCompilation.isReady &&
      clientCompilations.every(
        (compilationStatus) => compilationStatus.isReady
      );
    const isRendererReady = () => isBuildReady() && renderer;

    const renderQueue: Array<() => void> = [];
    const flushRenderQueue = async () => {
      if (isRendererReady() && renderQueue.length > 0) {
        await renderQueue.shift()!();
        flushRenderQueue();
      }
    };

    const render: Render<Route> = async (route: Route) => {
      const startRenderTime = Date.now();
      log(`Starting render`, route);
      const webpackStats = getClientStats();
      const renderParams = {
        ...route,
        ...mapStatsToParams({
          ...route,
          webpackStats,
        }),
      };
      try {
        const result = await renderer(renderParams);

        log(
          `Successfully rendered ${route.route} (${timeSince(startRenderTime)})`
        );
        return result;
      } catch (error) {
        error.webpackStats = webpackStats;
        throw error;
      }
    };

    const onRenderAll = async (currentCompilation: compilation.Compilation) => {
      log(`Starting routes render`);

      if (!rendererCompilation.compilation) {
        const errorMessage = `Unable to find render compilation. Something may have gone wrong during render build.`;
        logError(errorMessage);
        throw new Error(errorMessage);
      }

      try {
        await renderRoutes({
          render,
          renderConcurrency,
          renderCompilation: rendererCompilation.compilation,
          renderDirectory,
          renderEntry,
          routes,
          transformFilePath,
        });
      } catch (error) {
        logError("An error occured rendering HTML", error);
        currentCompilation.errors.push(new RenderError(error));
      }
      log(`Ending routes render`);
    };

    const getRenderEntry = (compilation: compilation.Compilation) => {
      const renderStats = compilation.getStats().toJson();

      const asset = renderStats.assetsByChunkName![renderEntry];
      if (!asset) {
        throw new Error(
          `Unable to find renderEntry "${renderEntry}" in assets. Possible entries are: ${Object.keys(
            renderStats.assetsByChunkName!
          ).join(", ")}.`
        );
      }

      let renderFile: any = asset;
      if (Array.isArray(renderFile)) {
        renderFile = renderFile[0];
      }
      if (renderFile && typeof renderFile === "object") {
        renderFile = renderFile.name;
      }

      return renderFile as string;
    };

    const renderCallbacks: any[] = [];

    const getClientStats = () => {
      if (lastClientStats) {
        return lastClientStats;
      }
      const clientStats =
        clientCompilations.length === 1
          ? clientCompilations[0].compilation!.getStats()
          : new MultiStats(
              clientCompilations
                .map((compilationStatus) => compilationStatus.compilation)
                .filter(Boolean)
                .map((compilation) => compilation!.getStats())
            );

      lastClientStats = clientStats;
      return clientStats;
    };

    const flushQueuedRenders = () => {
      if (isRendererReady() && renderCallbacks.length > 0) {
        renderCallbacks.shift()(renderer, getClientStats());
        flushQueuedRenders();
      }
    };

    const onRendererReady: OnRendererReady<Route> = (cb) => {
      if (isRendererReady()) {
        cb(render);
      } else {
        renderCallbacks.push(cb);
      }
    };

    const createRendererIfReady = async (
      currentCompilation: compilation.Compilation
    ) => {
      if (!isBuildReady()) {
        return;
      }
      const renderCompilation = rendererCompilation.compilation!;
      const renderEntry = getRenderEntry(renderCompilation);
      log("Render route:", { renderEntry });
      const source = getSourceFromCompilation(renderCompilation);
      renderer = createRenderer({
        source,
        fileName: renderEntry,
        extraGlobals,
      });

      if (typeof renderer !== "function") {
        throw new Error(
          `Unable to find render function. File "${renderEntry}". Received ${typeof renderer}.`
        );
      }
      flushQueuedRenders();
      flushRenderQueue();
      if (!skipAssets) {
        await onRenderAll(currentCompilation);
      }
    };

    const apply = (compiler: Compiler, isRenderer: boolean) => {
      const compilerName = compiler.name || compiler.options.name;

      const compilationStatus: CompilationStatus = {
        compilation: null,
        isReady: false,
      };

      if (isRenderer) {
        log(`Received render compiler: ${compilerName}`);
      } else {
        log(`Received compiler: ${compilerName}`);
      }

      if (isRenderer) {
        if (rendererCompilation) {
          throw new Error("Error. Unable to apply a second renderer");
        }
        rendererCompilation = compilationStatus;
      } else {
        clientCompilations.push(compilationStatus);
      }

      compiler.hooks.watchRun.tap(pluginName, () => {
        log(`Build started for for ${compilerName}.`);
        compilationStatus.isReady = false;
      });

      compiler.hooks.afterEmit.tapPromise(pluginName, async (compilation) => {
        log(`Assets emitted for ${compilerName}.`);
        compilationStatus.compilation = compilation;
        lastClientStats = null;
        compilationStatus.isReady = true;
        try {
          await createRendererIfReady(compilation);
        } catch (error) {
          compilation.errors.push(error);
        }
      });
    };
    this.statsCollectorPlugin = (compiler: Compiler) => apply(compiler, false);
    this.rendererPlugin = (compiler?: Compiler) => {
      // Support legacy behaviour of calling '.render()' until next breaking change
      if (!compiler) {
        console.warn(
          "Warning. Calling render nolonger required. Change htmlRenderPlugin.render() to htmlRenderPlugin.render"
        );
        return (compiler: Compiler) => apply(compiler, true);
      }
      return apply(compiler, true);
    };
    this.apply = (compiler: Compiler) => {
      console.warn(
        "Warning. Attempted to apply directly to webpack. Use htmlRenderPlugin.statsCollectorPlugin instead."
      );
      this.statsCollectorPlugin(compiler);
    };
    this.render = () => this.rendererPlugin;
    this.createDevRouter = () =>
      createDevRouter<Route>({
        transformExpressPath,
        getRouteFromRequest,
        onRendererReady,
        getClientStats,
        routes,
      });

    this.renderWhenReady = (route: Route) =>
      new Promise((resolve, reject) => {
        const onRender = () => {
          log("Rendering renderWhenReady onRender");
          try {
            resolve(render(route));
          } catch (error) {
            reject(error);
          }
        };
        if (isRendererReady()) {
          onRender();
        } else {
          renderQueue.push(onRender);
        }
      });
  }
  renderWhenReady: (route: Route) => Promise<string>;
  statsCollectorPlugin: (compiler: Compiler) => void;
  rendererPlugin: (compiler: Compiler) => void;
  render: () => (compiler: Compiler) => void;
  apply: (compiler: Compiler) => void;
  createDevRouter: () => Router;
};
