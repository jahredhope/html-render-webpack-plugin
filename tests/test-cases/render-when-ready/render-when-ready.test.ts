import webpack from "webpack";

import config from "./webpack.config";
import HtmlRenderPlugin from "../../../src";
import { BaseRoute } from "../../../src/common-types";
import { createInMemoryFileSystem } from "../../utils/memory-fs";

describe("renderWhenReady", () => {
  interface Route extends BaseRoute {
    route: string;
    language: string;
    forceError?: boolean;
  }

  it("should render a HTML once resolved", async (done) => {
    const routes: Route[] = [{ route: "/", language: "en-au" }];
    const htmlRenderPlugin = new HtmlRenderPlugin<Route>({
      mapStatsToParams: () => ({}),
      skipAssets: true,
      routes,
    });

    const promise = htmlRenderPlugin.renderWhenReady({
      route: "/new",
      language: "en-us",
    });

    const compiler = webpack(config(htmlRenderPlugin));

    const memoryFs = createInMemoryFileSystem();
    // @ts-ignore: Yes outputFileSystem does exist on MultiCompiler
    compiler.outputFileSystem = memoryFs;

    compiler.run(async () => {
      await expect(promise).resolves.toMatchSnapshot();
      done();
    });
  });

  it("should contain webpackStats when render errors", async (done) => {
    const routes: Route[] = [{ route: "/", language: "en-au" }];
    const htmlRenderPlugin = new HtmlRenderPlugin<Route>({
      mapStatsToParams: () => ({}),
      skipAssets: true,
      routes,
    });

    const promise = htmlRenderPlugin.renderWhenReady({
      route: "/new",
      language: "en-us",
      forceError: true,
    });

    const compiler = webpack(config(htmlRenderPlugin));

    const memoryFs = createInMemoryFileSystem();
    // @ts-ignore: Yes outputFileSystem does exist on MultiCompiler
    compiler.outputFileSystem = memoryFs;

    compiler.run(async () => {
      await expect(promise).rejects.toEqual(
        expect.objectContaining({ webpackStats: expect.any(Object) })
      );
      done();
    });
  });
});
