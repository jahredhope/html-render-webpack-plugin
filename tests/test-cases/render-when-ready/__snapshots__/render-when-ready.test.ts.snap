// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderWhenReady should render a HTML once resolved 1`] = `
"<html>
<body>
  Rendered with:&nbsp;
  <code>[
    {
      \\"route\\": \\"/new\\",
      \\"language\\": \\"en-us\\"
    }
  ]</code>
</body>
</html>"
`;
